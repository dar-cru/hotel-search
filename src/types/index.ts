export * from './hotel';
export * from './offer';
export * from './property';
export * from './rating';
