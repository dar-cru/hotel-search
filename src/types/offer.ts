export enum PromotionType {
  CAMPAIGN = 'CAMPAIGN',
  MEMBER = 'MEMBER'
}

export enum CancellationType {
  NOT_REFUNDABLE = 'NOT_REFUNDABLE',
  FREE_CANCELLATION = 'FREE_CANCELLATION'
}

export type Offer = {
  promotion: {
    title: string;
    type: PromotionType;
  };
  name: string;
  displayPrice: {
    amount: number;
    currency: string;
  };
  savings?: {
    amount: number;
    currency: string;
  };
  cancellationOption: {
    cancellationType: CancellationType;
  };
};
