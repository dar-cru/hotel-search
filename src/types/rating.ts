export enum RatingType {
  SELF = 'self',
  STAR = 'star'
}

export type Rating = {
  ratingValue: 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5;
  ratingType: RatingType;
};
