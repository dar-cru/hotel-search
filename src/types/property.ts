import { Rating } from './rating';

export type Property = {
  propertyId: string;
  title: string;
  address: string[];
  previewImage: {
    url: string;
    caption: string;
    imageType: string;
  };
  rating: Rating;
};
