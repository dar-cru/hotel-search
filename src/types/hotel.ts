import { Offer } from './offer';
import { Property } from './property';

export type Hotel = {
  id: string;
  property: Property;
  offer: Offer;
};
