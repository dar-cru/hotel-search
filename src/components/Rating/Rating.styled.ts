import styled from 'styled-components';
import { RatingType } from '../../types';

export const StarsRating = styled.div<{ ratingType: RatingType; rating: number }>`
  display: inline-block;
  font-size: ${(props) => (props.ratingType === RatingType.STAR ? '1rem' : '1.5rem')};
  font-family: Times;
  line-height: 1;

  :before {
    ${(props) => (props.ratingType === RatingType.STAR ? `content: '★★★★★'` : `content: '●●●●●'`)};
    letter-spacing: ${(props) => (props.ratingType === RatingType.STAR ? '3px' : '1px')};;
    background: linear-gradient(90deg, gold ${(props) => props.rating}%, grey ${(props) => props.rating}%);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
`;
