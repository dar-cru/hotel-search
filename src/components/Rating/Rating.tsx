import { RatingType } from '../../types/rating';
import { StarsRating } from './Rating.styled';

export type RatingProps = {
  ratingType: RatingType;
  ratingValue: 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5;
};

const Rating = ({ ratingType, ratingValue }: RatingProps) => {
  const percentageRating = (ratingValue / 5) * 100;
  return <StarsRating ratingType={ratingType} rating={percentageRating} />;
};

export default Rating;
