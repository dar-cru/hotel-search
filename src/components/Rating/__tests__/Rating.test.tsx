import { render, RenderResult } from '@testing-library/react';
import { RatingType } from '../../../types';
import Rating, { RatingProps } from '../Rating';

describe('Rating component', () => {
  it('renders the component with star rating', () => {
    const props: RatingProps = {
      ratingType: RatingType.STAR,
      ratingValue: 4
    };
    const tree: RenderResult = render(<Rating {...props} />);

    const { container } = tree;
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule('content', `'★★★★★'`, { modifier: ':before' });
    expect(container.firstChild).toHaveStyleRule('background', 'linear-gradient(90deg,gold 80%,grey 80%)', {
      modifier: ':before'
    });
  });

  it.each`
    rating | gradientPercentage
    ${1}   | ${'20'}
    ${1.5} | ${'30'}
    ${2}   | ${'40'}
    ${2.5} | ${'50'}
    ${3}   | ${'60'}
    ${3.5} | ${'70'}
    ${4}   | ${'80'}
    ${4.5} | ${'90'}
    ${5}   | ${'100'}
  `('renders the correct background gradient for each star rating value', ({ rating, gradientPercentage }) => {
    const props: RatingProps = {
      ratingType: RatingType.STAR,
      ratingValue: rating
    };
    const { container }: RenderResult = render(<Rating {...props} />);
    expect(container.firstChild).toHaveStyleRule('content', `'★★★★★'`, { modifier: ':before' });
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `linear-gradient(90deg,gold ${gradientPercentage}%,grey ${gradientPercentage}%)`,
      {
        modifier: ':before'
      }
    );
  });

  it('renders the component with self rating', () => {
    const props: RatingProps = {
      ratingType: RatingType.SELF,
      ratingValue: 5
    };
    const tree: RenderResult = render(<Rating {...props} />);

    const { container } = tree;
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule('content', `'●●●●●'`, { modifier: ':before' });
    expect(container.firstChild).toHaveStyleRule('background', 'linear-gradient(90deg,gold 100%,grey 100%)', {
      modifier: ':before'
    });
  });

  it.each`
    rating | gradientPercentage
    ${1}   | ${'20'}
    ${1.5} | ${'30'}
    ${2}   | ${'40'}
    ${2.5} | ${'50'}
    ${3}   | ${'60'}
    ${3.5} | ${'70'}
    ${4}   | ${'80'}
    ${4.5} | ${'90'}
    ${5}   | ${'100'}
  `('renders the correct background gradient for each self rating value', ({ rating, gradientPercentage }) => {
    const props: RatingProps = {
      ratingType: RatingType.SELF,
      ratingValue: rating
    };
    const { container }: RenderResult = render(<Rating {...props} />);
    expect(container.firstChild).toHaveStyleRule('content', `'●●●●●'`, { modifier: ':before' });
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `linear-gradient(90deg,gold ${gradientPercentage}%,grey ${gradientPercentage}%)`,
      {
        modifier: ':before'
      }
    );
  });
});
