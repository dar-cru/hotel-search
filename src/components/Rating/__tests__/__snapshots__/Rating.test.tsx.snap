// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating component renders the component with self rating 1`] = `
.c0 {
  display: inline-block;
  font-size: 1.5rem;
  font-family: Times;
  line-height: 1;
}

.c0:before {
  content: '●●●●●';
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  background: linear-gradient(90deg,gold 100%,grey 100%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Rating component renders the component with star rating 1`] = `
.c0 {
  display: inline-block;
  font-size: 1rem;
  font-family: Times;
  line-height: 1;
}

.c0:before {
  content: '★★★★★';
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
  background: linear-gradient(90deg,gold 80%,grey 80%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

<div>
  <div
    class="c0"
  />
</div>
`;
